### ISSUE TITLE
Serializing typed Lists and Dicts?

### ISSUE THREAD

**Issue by JackMcCoy** [numba/numba#4698]
I have preprocessed a large amount of data (Levenshtein matches, array locations, etc.) that I've stored as python object pickles. When my program runs, it performs operations on a subset of that data, and first goes through and pulls out the data needed by the later processing, gets the intersections of sets, and things like that. I've sped the process up a bit using a numba njit method, but would like to 1) be able to run that method in parallel processes and 2) not have to import the python objects/loop through them each time the program runs to build the numba objects.

Multiprocessing can be accomplished using a python method as an intermediary between the pool and numba, but I am running this on Windows at the moment, which has spawned, rather than forked processes. Trying to read four copies of the pickled objects in/building the nested numba Lists and Dicts crashes my system.

Specifics of the program aside, how can you serialize, say, an object of type DictType[unicode_type,ListType[unicode_type]] to reduce some of this cost? Trying to use pickle obviously results in an error that MemInfo can't be pickled, and I've searched the source code for hours, seeing things like, boxing and unboxing, and reduce() but haven't made much headway figuring out what parameters like (typ, val, c) mean. Can anyone help?

**Comment by stuartarchibald** [comment]
Thanks for the request. What you are asking for isn't possible right now, there's Numba internal memory in use in the List/Dict instances which does not pickle. Adding this as a feature request.

**Comment by esc** [comment]
For the typed-list it might be feasible to re-interpret the memory used as a NumPy array and then use one of the common serialization methods for NumPy arrays.

**Comment by mstewart141** [comment]
first off, thanks to the numba team for the great library.

i believe this would be a valuable addition: my usecase is that i have a feature transformer (think e.g. sklearn's `StandardScalar` that needs to be fit on a dataset, and is later used to transform another dataset

```
xformer = Xformer()
xformer.fit(data)
xformer.transform(data2)
```

the `fit` method attaches a `numba` typed dict to `self` which is then used for the `transform` step.

currently there is no workaround except to copy it over to a python dictionary and then basically recopy it back into `numba` every time i deserialize which is not ideal (but works).

warm regards

**Comment by esc** [comment]
@mstewart141 thanks for commenting on this issue and the praise you offer, we do appreciate it! Interestingly enough, all previous conversations about this topic ended up concluding that the simplest solution would be to write the pickler so that it serializes the typed Numba object as a Python object and deserializes from that too, which is what you have already.

**Comment by mstewart141** [comment]
@esc i report back to you now in something of a bind haha. the situation is this:

i had to switch to a 64 bit hash function for more entropy (i am operating w/in a `numba.njit` function on hashed strings, in a large numpy uint array a la `uint64[:, :]`) meaning that my `typed Dict` now has `uint64` keys. this is all well and good (my transformer can both fit and predict successfully, as before).

_however_, _however_, there is a catch: previously to serialize my `numba` dict i was simply calling dict() on it -- which worked great. now, though, so much as calling `d.items()` crashes with a trace as 

```
safe.py:64: in make_serializable
    self.nested_map_ = {idx: dict(inner) for idx, inner in self.nested_map_.items()}
safe.py:64: in <dictcomp>
    self.nested_map_ = {idx: dict(inner) for idx, inner in self.nested_map_.items()}
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[OverflowError("int too big to convert") raised in repr()] Dict object at 0x7f4f778cc4e0>
key = 10669984055021876834

    def __getitem__(self, key):
        if not self._typed:
            raise KeyError(key)
        else:
>           return _getitem(self, key)
E           OverflowError: int too big to convert

key        = 10669984055021876834
self       = <[OverflowError("int too big to convert") raised in repr()] Dict object at 0x7f4f778cc4e0>
```

the `key` is certainly a uint64. somewhere, however, there appears to be support only for values in the signed int64 range. do let me know if you happen to have any leads, and thanks again!

**Comment by androbaza** [comment]
is this still not possible? i see that NumbaPickler exists in numba.core.serialize, but if I try to use it as pickler for joblib Parallel, it still gives out an error/

**Comment by esc** [comment]
> is this still not possible? i see that NumbaPickler exists in numba.core.serialize, but if I try to use it as pickler for joblib Parallel, it still gives out an error/

Yes, AFAICT, it is still not possible and currently not scheduled for implementation, as far as I know.
